File: py/test_md.py
"""openctp TTS 行情连接测试"""
import fix_locale  # noqa: F401  must precede openctp_ctp
import os
import time
from pathlib import Path
from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parent.parent / ".env")

from openctp_ctp import mdapi

BROKER_ID = os.getenv("CTP_BROKER_ID")
USER_ID = os.getenv("CTP_USER_ID")
PASSWORD = os.getenv("CTP_PASSWORD")
MD_FRONT = os.getenv("CTP_MD_FRONT")

SUBSCRIBE_INSTRUMENTS = ["ag2606", "IF2604"]


class MdSpi(mdapi.CThostFtdcMdSpi):
    def __init__(self, api):
        super().__init__()
        self.api = api
        self.login_ok = False
        self.tick_count = 0

    def OnFrontConnected(self):
        print("[MD] 前置连接成功")
        req = mdapi.CThostFtdcReqUserLoginField()
        req.BrokerID = BROKER_ID
        req.UserID = USER_ID
        req.Password = PASSWORD
        self.api.ReqUserLogin(req, 0)

    def OnFrontDisconnected(self, nReason):
        print(f"[MD] 前置断开连接, 原因: {nReason}")

    def OnRspUserLogin(self, pRspUserLogin, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(f"[MD] 登录失败: {pRspInfo.ErrorID} - {pRspInfo.ErrorMsg}")
            return
        self.login_ok = True
        print(f"[MD] 登录成功, 交易日: {pRspUserLogin.TradingDay}")
        print(f"[MD] 订阅合约: {SUBSCRIBE_INSTRUMENTS}")
        self.api.SubscribeMarketData(
            [i.encode() if isinstance(i, str) else i for i in SUBSCRIBE_INSTRUMENTS],
            len(SUBSCRIBE_INSTRUMENTS),
        )

    def OnRspSubMarketData(self, pSpecificInstrument, pRspInfo, nRequestID, bIsLast):
        if pRspInfo and pRspInfo.ErrorID != 0:
            print(
                f"[MD] 订阅失败 {pSpecificInstrument.InstrumentID}: "
                f"{pRspInfo.ErrorMsg}"
            )
        else:
            print(f"[MD] 订阅成功: {pSpecificInstrument.InstrumentID}")

    def OnRtnDepthMarketData(self, pData):
        self.tick_count += 1
        print(
            f"[MD] {pData.InstrumentID:>10} | "
            f"最新: {pData.LastPrice:>10.2f} | "
            f"买一: {pData.BidPrice1:>10.2f} x {pData.BidVolume1:<5} | "
            f"卖一: {pData.AskPrice1:>10.2f} x {pData.AskVolume1:<5} | "
            f"成交量: {pData.Volume}"
        )
        if self.tick_count >= 20:
            print(f"\n[MD] 已收到 {self.tick_count} 个 tick，行情测试通过")


def run_md_test():
    print("=" * 60)
    print("openctp TTS 行情连接测试")
    print(f"BrokerID: {BROKER_ID}, UserID: {USER_ID}")
    print(f"MD Front: {MD_FRONT}")
    print(f"订阅合约: {SUBSCRIBE_INSTRUMENTS}")
    print("=" * 60)

    api = mdapi.CThostFtdcMdApi.CreateFtdcMdApi("./md_flow/")
    spi = MdSpi(api)
    api.RegisterSpi(spi)
    api.RegisterFront(MD_FRONT)
    api.Init()

    for _ in range(30):
        time.sleep(1)
        if spi.tick_count >= 20:
            break

    if not spi.login_ok:
        print("[MD] 超时: 未能成功登录")
    elif spi.tick_count == 0:
        print("[MD] 登录成功但未收到行情（可能非交易时间）")
    else:
        print(f"[MD] 测试完成，共收到 {spi.tick_count} 个 tick")
    return spi.login_ok


def test_md_connectivity():
    """行情接口连通性测试"""
    assert run_md_test(), "MD login failed within timeout"


if __name__ == "__main__":
    import sys
    sys.exit(0 if run_md_test() else 1)

File: cpp/test_md.cpp
#include <cstdio>
#include <cstring>
#include <string>
#include <thread>
#include <chrono>
#include <atomic>
#include <mutex>
#include <condition_variable>
#include "ThostFtdcMdApi.h"
#include "env_loader.h"
#include "fix_locale.h"

static std::string BROKER_ID, USER_ID, PASSWORD, MD_FRONT;
static std::atomic<bool> g_login_ok{false};
static std::atomic<int> g_tick_count{0};
static std::mutex g_mtx;
static std::condition_variable g_cv;

static const char *INSTRUMENTS[] = {"ag2606", "IF2604"};
static const int NUM_INSTRUMENTS = 2;

class MdSpi : public CThostFtdcMdSpi {
public:
    explicit MdSpi(CThostFtdcMdApi *api) : api_(api) {}

    void OnFrontConnected() override {
        printf("[MD] Connected\n");
        CThostFtdcReqUserLoginField req{};
        strncpy(req.BrokerID, BROKER_ID.c_str(), sizeof(req.BrokerID) - 1);
        strncpy(req.UserID, USER_ID.c_str(), sizeof(req.UserID) - 1);
        strncpy(req.Password, PASSWORD.c_str(), sizeof(req.Password) - 1);
        api_->ReqUserLogin(&req, 0);
    }

    void OnFrontDisconnected(int nReason) override {
        printf("[MD] Disconnected, reason: %d\n", nReason);
    }

    void OnRspUserLogin(CThostFtdcRspUserLoginField *pRspUserLogin, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0) {
            printf("[MD] Login failed: %d - %s\n", pRspInfo->ErrorID, pRspInfo->ErrorMsg);
            return;
        }
        g_login_ok = true;
        printf("[MD] Login OK, TradingDay: %s\n", pRspUserLogin->TradingDay);

        char *ppInstrumentID[NUM_INSTRUMENTS];
        for (int i = 0; i < NUM_INSTRUMENTS; i++)
            ppInstrumentID[i] = const_cast<char *>(INSTRUMENTS[i]);
        api_->SubscribeMarketData(ppInstrumentID, NUM_INSTRUMENTS);
    }

    void OnRspSubMarketData(CThostFtdcSpecificInstrumentField *pInstrument, CThostFtdcRspInfoField *pRspInfo, int nRequestID, bool bIsLast) override {
        if (pRspInfo && pRspInfo->ErrorID != 0)
            printf("[MD] Subscribe failed %s: %s\n", pInstrument->InstrumentID, pRspInfo->ErrorMsg);
        else
            printf("[MD] Subscribed: %s\n", pInstrument->InstrumentID);
    }

    void OnRtnDepthMarketData(CThostFtdcDepthMarketDataField *pData) override {
        int cnt = ++g_tick_count;
        printf("[MD] %10s | Last: %10.2f | Bid: %10.2f x %-5d | Ask: %10.2f x %-5d | Vol: %d\n",
               pData->InstrumentID, pData->LastPrice,
               pData->BidPrice1, pData->BidVolume1,
               pData->AskPrice1, pData->AskVolume1,
               pData->Volume);
        if (cnt >= 20) {
            printf("\n[MD] Received %d ticks, test passed\n", cnt);
            g_cv.notify_all();
        }
    }

private:
    CThostFtdcMdApi *api_;
};

int main() {
    fix_locale();
    auto env = load_env("../.env");
    BROKER_ID = env["CTP_BROKER_ID"];
    USER_ID   = env["CTP_USER_ID"];
    PASSWORD  = env["CTP_PASSWORD"];
    MD_FRONT  = env["CTP_MD_FRONT"];

    printf("============================================================\n");
    printf("openctp TTS MD connectivity test (C++)\n");
    printf("BrokerID: %s, UserID: %s\n", BROKER_ID.c_str(), USER_ID.c_str());
    printf("MD Front: %s\n", MD_FRONT.c_str());
    printf("============================================================\n");

    auto *api = CThostFtdcMdApi::CreateFtdcMdApi("./md_flow/");
    MdSpi spi(api);
    api->RegisterSpi(&spi);
    api->RegisterFront(const_cast<char *>(MD_FRONT.c_str()));
    api->Init();

    {
        std::unique_lock<std::mutex> lk(g_mtx);
        if (!g_cv.wait_for(lk, std::chrono::seconds(30), [] { return g_tick_count >= 20; })) {
            if (!g_login_ok) {
                printf("[MD] Timeout: login failed\n");
                api->RegisterSpi(nullptr);
                api->Release();
                return 1;
            }
            if (g_tick_count == 0) {
                printf("[MD] Login OK but no ticks (may not be trading hours)\n");
            }
        }
    }

    printf("[MD] Test completed, received %d ticks\n", g_tick_count.load());
    api->RegisterSpi(nullptr);
    api->Release();
    return 0;
}
